grpc/noncebalancer/noncebalancer.go

```diff
@@ -0,0 +1,84 @@
+package noncebalancer
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/letsencrypt/boulder/nonce"
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/balancer/base"
+)
+
+// Compile-time assertion that *Picker implements the base.PickerBuilder
+// interface.
+var _ base.PickerBuilder = (*Picker)(nil)
+
+// Picker is a base.Balancer used to construct a balancer.Picker capable of
+// picking a backend (SubConn) based on the context of each RPC message. It
+// implements the base.PickerBuilder and balancer.Picker interfaces but should
+// only be used as a base.Balancer for nonce server clients. It can be invoked
+// by passing `{"loadBalancingConfig":[{"nonce":{}}]}` as the default service
+// config to grpc.Dial().
+type Picker struct {
+	sync.RWMutex
+	backends map[balancer.SubConn]base.SubConnInfo
+}
+
+// Build implements the base.PickerBuilder interface. It is called by the gRPC
+// runtime when the balancer is first initialized and when the set of backend
+// (SubConn) addresses changes. It is responsible for initializing the Picker's
+// backends map and returning a balancer.Picker.
+func (p *Picker) Build(buildInfo base.PickerBuildInfo) balancer.Picker {
+	p.Lock()
+	defer p.Unlock()
+	p.backends = buildInfo.ReadySCs
+	return p
+}
+
+// Compile-time assertion that *Picker implements the balancer.Picker interface.
+var _ balancer.Picker = (*Picker)(nil)
+
+// Pick implements the balancer.Picker interface. It is called by the gRPC
+// runtime for each RPC message. It is responsible for picking a backend
+// (SubConn) based on the context of each RPC message.
+func (p *Picker) Pick(info balancer.PickInfo) (balancer.PickResult, error) {
+	p.RLock()
+	defer p.RUnlock()
+
+	var result balancer.PickResult
+	if len(p.backends) == 0 {
+		// The Picker must be rebuilt if there are no backends available.
+		return result, balancer.ErrNoSubConnAvailable
+	}
+
+	// Get the destination prefix from the RPC context.
+	destPrefix := info.Ctx.Value(nonce.PrefixKey{}).(string)
+	if destPrefix == "" {
+		// This should never happen.
+		return result, errors.New("'prefix' value required in RPC context")
+	}
+
+	// Get the salt from the RPC context.
+	salt := info.Ctx.Value(nonce.PrefixSaltKey{}).(string)
+	if salt == "" {
+		// This should never happen.
+		return result, errors.New("'salt' value required in RPC context")
+	}
+
+	// Iterate over the backends and return the first one that matches the
+	// destination prefix from the RPC context.
+	for sc, scInfo := range p.backends {
+		scPrefix := nonce.DerivePrefix(scInfo.Address.Addr, salt)
+		if scPrefix == destPrefix {
+			result.SubConn = sc
+			return result, nil
+		}
+	}
```

Comment:
I now understand why we can't do this at `.Build()` time: because we need the salt, and the rpc context is the only way to smuggle the salt into here.

But I think we can still do this only once by memoizing it -- store a map of scPrefix to SubConn on the Picker struct, nil that map inside Build(), and only do this loop here if that map is nil (and then save the result back to the map).

Reply:
I thought about doing exactly this. But it would require obtaining a RW lock within `Pick()`, which would be fine, but `Pick()` isn't supposed to block per:
https://pkg.go.dev/google.golang.org/grpc/balancer#Picker

Reply:
A few thoughts:
1) If we think that computing the mapping of prefixes to subconns is "time-consuming work", then we should definitely try to only do it in one Pick() call rather than all of them
2) If instead we're concerned that getting a write-lock to update the p.backendMap (or whatever we call it) value is a bad thing to do, we can just not do that: updating the map is an idempotent operation because every call to Pick() should result in computing the same map, so it's okay to let multiple Pick() calls all write to it at the same time
3) If we're really concerned about that anyway, because maybe the value of `p.backends` is being updated really rapidly and so multiple simultaneous Pick() calls are building maps from different inputs, then let's actually split the Balancer and the Picker into two separate structs and have the Build() method actually return a new object rather than just returning itself. Then each Picker that it returns will never have its own `p.backends` updated (because the Builder will just be producing a brand new Picker instead) and writing to the `p.backendMap` becomes truly idempotent.

Reply:
Thanks for pointing this out 👍🏻. I believe that option 3 sounds like the safest approach.